fix issue with slang version string

piotr-binkowski opened the pull request:
using tag name and number of commits since that tag as version creates issues when tags in the repository are not getting sorted properly ie. conda sees nightly as lower version compared to v0.1
as can be seen [here](https://anaconda.org/SymbiFlow/slang). ~Using date of the last commit seems to be the most reliable solution.~ As pointed out in SymbiFlow/sv-tests#633 this is currently an issue in sv-tests.

Commit:
add filtered tag names for use in version strings

Signed-off-by: Piotr Binkowski <pbinkowski@antmicro.com>

piotr-binkowski commented:
I've modified the scripts a bit to generate git tag and commit count variables manually, with that I can filter out the `nightly` tag and use the previous one.

tgorochowik reviewed (commented):
We've had this discussion some time ago, I think that @mithro said that there is a way to make conda a little bit smarter about the tags (so it treats version tags differently).

If not then the system we use is broken, and I am afraid your current patch does not scale.

If there is no better solution then I think we should white-list tags like: `v[0-9]\.[0-9]-?.*` (or whatever works) and use it for all packages. Blacklisting will need to be maintained manually and the packages will be breaking every now and then.

piotr-binkowski commented:
I see that there is an issue conda/conda-build#3818 by @mithro which suggests to add support for tag filtering but from what I see there are no responses and the issue is still open so probably this is not implemented. I can modify this to allow tag whitelisting for all packages if that is the acceptable solution.

tgorochowik commented:
I think that makes sense, we should solve this somehow here until conda/conda-build#3818 is resolved.

I guess you could use the `--match` parameter mentioned there.

In my opinion it is also a good moment to get rid of the requirement to actually have any matching tags in the repo (like the `v0.0` that is added to some of the repositories right now). If we are customizing it completely we could just add `v0.0` manually if no matching tag is found.

piotr-binkowski commented:
I've updated the scripts to now use the `git describe` with `--match` option to match only tags that we are interested in. I've also modified `meta.yaml` of packages where git tags were compatible with the default pattern to use new variables.

tgorochowik commented on .travis/tag-filter.sh:
```diff
@@ -0,0 +1,34 @@
+TAG_GIT_URL=$(cat $PACKAGE/meta.yaml | grep "git_url" | awk '{print $2}')
+TAG_GIT_REV=$(cat $PACKAGE/meta.yaml | grep "git_rev" | awk '{print $2}')
+TAG_PATTERN=${TAG_PATTERN:-"v[0-9]*\.[0-9]*"}
```
Should we allow any annotations after the actual version? For example `v0.1-some-note`.

BTW. How are these be sorted in the conda repository? Are we sure that `v2.0` and `v10.0` will get sorted correctly (i.e not alphabetically)?

piotr-binkowski replied:
From what I see [conda-build documentation](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#package-version) recommends using [PEP 386](https://www.python.org/dev/peps/pep-0386/) but I'm not sure if its going to handle the leading `v` correctly.
Patterns used by `git describe` seem to follow `tcl glob` pattern syntax which means that `[0-9]*` are in fact 2 separate patterns `[0-9]` and `.*`. As you can see [here](https://travis-ci.com/SymbiFlow/conda-packages/jobs/285875635#L284) it already matches tags with extra information

tgorochowik replied:
Sounds good to me. I think we can assume that we are only willing to parse tags starting with `v` and fall back to `v0.0` if the repo does something else.

Let @mithro take a look at that before we merge it though.

mithro replied:
You need to strip the leading 'v'.

mithro commented on .travis/tag-filter.sh:
```diff
@@ -0,0 +1,34 @@
+TAG_PATTERN=${TAG_PATTERN:-"v[0-9]*\.[0-9]*"}
+TAG_TMP_DIR=$(mktemp -d)
+
+git clone -b "$TAG_GIT_REV" "$TAG_GIT_URL" $TAG_TMP_DIR && pushd $TAG_TMP_DIR
```
Conda will download the git repository (in "bare" mode) to `conda/conda-bld/git_cache/github.com/SymbiFlow/yosys.git` -- I suggest we reuse this location so we don't end up downloading the git repository multiple times.

With a bare repository you set `GIT_DIR` to the correct location. IE
```
GIT_DIR=~/conda/conda-bld/git_cache/github.com/SymbiFlow/yosys.git git describe --long --tags
```
This also should be part of `conda-env.sh` script.

piotr-binkowski replied:
I've modified the script to use `.../conda-bld/git_cache` directories but I'm not sure how should I make this part of `conda-env.sh`. I've added a `source .travis/tag-filter.sh` there but now it requires `PACKAGE` and `CONDA_PATH` to be set.

Commit:
add tag-filter to conda-env.sh

Signed-off-by: Piotr Binkowski <pbinkowski@antmicro.com>

hzeller commented:
Is this ready to be submitted @mithro ? It would unblock non-building verible as well.

mithro committed:
scripts: Add conda to PATH in conda-env.sh

(Also check that conda exists.)

Signed-off-by: Tim 'mithro' Ansell <me@mith.ro>

mithro commented:
@piotr-binkowski I created https://github.com/SymbiFlow/conda-packages/pull/74 based on your ideas but alternative implementation to this pull request. Can you take a look at it and see if it works for you?

It also generally tries to make it closer between what happens on your local machine and what happens on Travis.
